### PR TITLE
Fixing thread pool executor

### PR DESCRIPTION
**Pull request by hkaiser**
This patch fixes the problems with the thread pool executors. It also includes 2 fly-by changes

- to the shared state of `hpx::future`, fixing a benign data race
- to the promise test

It also includes a new test and an extension to an existing test

### CODE REVIEW

**Comment by hkaiser** [comment]
I'll go ahead an merge this PR first as it makes most of buildbot green. It will enable us to better see the impact of further PRs on the builds.

File: hpx/runtime/threads/executors/thread_pool_executors.hpp [hunk 1]
```diff
@@ -104,6 +107,13 @@ namespace hpx { namespace threads { namespace executors
 
             // resource manager registration
             std::size_t cookie_;
+
+            // store the self reference to the HPX running this scheduler
```
Comment (sithhell) [inline hpx/runtime/threads/executors/thread_pool_executors.hpp]: typo in the comment here: "to the HPX threads"

File: hpx/state.hpp [hunk 2]
```diff
@@ -24,11 +24,12 @@ namespace hpx
         state_pre_main = 3,
         state_starting = 4,
         state_running = 5,
-        state_pre_shutdown = 6,
-        state_shutdown = 7,
-        state_stopping    = 8,
-        state_terminating = 9,
-        state_stopped = 10,
+        state_suspended = 6,
```
Comment (sithhell) [inline hpx/state.hpp]: What does it mean when the runtime is suspended?

File: src/util/command_line_handling.cpp [hunk 3]
```diff
@@ -145,7 +145,8 @@ namespace hpx { namespace util
             std::size_t num_localities)
         {
             std::size_t batch_localities = env.retrieve_number_of_localities();
-            if (num_localities == 1)
+            if (num_localities == 1 &&
+                batch_localities != std::size_t(batch_localities))
```
Comment (sithhell) [inline src/util/command_line_handling.cpp]: This doesn't make a lot of sense to me. Isn't this always false? `batch_localities` already is of type std::size_t.

Comment (hkaiser) [inline src/util/command_line_handling.cpp]: Good catch, should be `batch_localities != std::size(-1)` - doh!

File: hpx/state.hpp (hunk 2 above)
Comment (hkaiser) [inline hpx/state.hpp]: We merged the states for the runtime and the thread manager. For the scheduler to be suspended makes sense for the thread pool executors only. Those 'suspend' by giving back control to the base-scheduler on a regular basis (see idle-callback, [here](https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/runtime/threads/detail/scheduling_loop.hpp#L421), and [here](https://github.com/STEllAR-GROUP/hpx/blob/master/src/runtime/threads/executors/thread_pool_executors.cpp#L277))

File: hpx/runtime/threads/executors/thread_pool_executors.hpp (hunk 1 above)
Comment (hkaiser) [inline hpx/runtime/threads/executors/thread_pool_executors.hpp]: Yes, thanks. I have that fixed already. There are more things to clean up from the latest test results.

File: src/util/command_line_handling.cpp (hunk 3 above)
Comment (hkaiser) [inline src/util/command_line_handling.cpp]: Please see #1574 for a related fix

**Pull request merged**